### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by banister**
Windows support. Requires `curl` and `7zip` to be installed, but `etags` is bundled with the gem.

### CODE REVIEW

**Comment by banister** [comment]
@kyrylo any feedback or shall i just merge?

File: lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          curl -k --fail -L -O https://github.com/ruby/ruby/archive/v#{ruby_version}.zip & 7z -y x v#{ruby_version}.zip
+        }
+      else
+        "curl --fail -L https://github.com/ruby/ruby/archive/v#{ruby_version}.tar.gz | tar xzvf - 2> /dev/null"
+      end
+    end
+
+    def etag_binary
+      @etag_binary ||= if Pry::Platform.linux?
+                         arch = RbConfig::CONFIG['arch'] =~ /i(3|6)86/ ? 32 : 64
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb]: `(3|6)` -> `[36]`

File: lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    def download_ruby
+      FileUtils.mkdir_p(ruby_source_folder)
+      FileUtils.cd(File.dirname(ruby_source_folder)) do
+        %x{ #{curl_cmd} }
+        check_for_error(curl_cmd) { Dir.entries(ruby_source_folder).count > 5 }
+      end
+    end
+
+    # @param [String] message Message to display on error
+    # @param [&Block] block Optional assertion
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb]: Not sure it's valid YARD.

File: lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    def check_for_error(message, &block)
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb]: The method name is super general and doesn't tell me anything.

File: lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      if Pry::Platform.windows?
+        %{dir /b /s *.c *.h *.y | "#{etag_binary}" - --no-members}
+      else
+        "find . -type f -name '*.[chy]' | #{etag_binary} - --no-members"
+      end
+    end
+
+    def generate_tagfile
+      FileUtils.cd(ruby_source_folder) do
+        puts "Generating tagfile!"
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb]: May be useful to spell out `ruby_source_folder` too.

File: lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb [hunk 5]
```diff
@@ -85,61 +88,10 @@ def self.parse_tagfile
...
-    end
-
-    def self.generate_tagfile
-      find_cmd = "find . -type f -name '*.[chy]' | #{etag_binary} - --no-members"
-
-      FileUtils.cd(ruby_source_folder) do
-        puts "Generating tagfile!"
-        %x{ #{find_cmd} }
-        check_for_error(find_cmd) { File.size(File.join(ruby_source_folder, "TAGS")) > 500 }
+      if !File.exists?(tags)
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: Idiomatic ruby would be `unless File.exists?`

File: lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb [hunk 6]
```diff
@@ -85,61 +88,10 @@ def self.parse_tagfile
...
+        ruby_source_installer.install
       end
+      @tagfile ||= File.read(tags)
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: Why do we need `||=` instead of `=`?

File: lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    def install
+      ask_for_install
+      puts "Downloading and setting up Ruby #{ruby_version} source..."
+      download_ruby
+      generate_tagfile
+      puts "...Finished!"
+    end
+
+    def ask_for_install
+      print "Identifier not found - do you want to install CRuby sources to attempt to resolve the identifier there?\nThis allows the lookup of C internals Y/N "
```
Comment (kyrylo) [inline lib/pry-doc/pry_ext/show_source_with_c_internals/ruby_source_installer.rb]: We could split long lines with backslashes (here and on line 23).

File: lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb (hunk 6 above)
Comment (banister) [reply lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: @kyrylo we don't technically, but might as well memoize as it's an expensive operation, and in the future another method might call it i guess

Comment (kyrylo) [reply lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: I think we might need to do something similar to what you did in https://github.com/pry/pry-doc/pull/85/commits/97498182940d07ecc213730a6cfa7e33a5332135 because we would be calculating previous lines nevertheless.

Comment (banister) [reply lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: @kyrylo how so? all the methods are memoized, including callers, an callees

Comment (kyrylo) [reply lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: It's brittle and if you change the code in the future it's not obvious there's such a requirement for this method. `File.exists?(tags)` isn't memoized.

All we need is to write a guard and change `||=` to `=`:

```rb
def self.tagfile
  return @tagfile if @tagfile
  # ...
  @tagfile = File.read(tags)
end
```

File: lib/pry-doc.rb [hunk 8]
```diff
@@ -23,4 +18,8 @@ def self.root
   root
 end
 
+require 'yard'
+require 'pry-doc/version'
+require 'pry-doc/pry_ext/method_info'
```
Comment (kyrylo) [inline lib/pry-doc.rb]: How is this change relevant to this PR and what's the motivation?

File: lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb (hunk 6 above)
Comment (banister) [reply lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb]: I was referring to this comment of yours: "because we would be calculating previous lines nevertheless." which i found pretty confusing :smile: But sure, i can fold `File.exists?` inside

File: lib/pry-doc.rb (hunk 8 above)
Comment (banister) [reply lib/pry-doc.rb]: @kyrylo i explain in the commit message https://github.com/pry/pry-doc/pull/85/commits/968b073ce9ea077e8640ebf613c176422c11f96c

```
require files after PryDoc class is configured

This is because we need to ensure PryDoc.root is defined before loading other classes that depend on it
```

Comment (banister) [reply lib/pry-doc.rb]: @kyrylo tests started to fail as i was setting up `RubySourceInstaller` in the class body, and `RubySourceInstaller` needed access to the `PryDoc.root` method - and all this happened BEFORE `PryDoc.root` was defined when the requires were above the definition, so i just moved the requires under

**Comment by kyrylo** [comment]
When I run tests locally, I get the following error:

```rb
Failures:

  1) PryDoc Pry::CInternals::CodeFetcher no tags file exists attempts to install and setup ruby
     Failure/Error: expect(described_class.ruby_source_installer).to receive(:install)

       (#<Pry::CInternals::RubySourceInstaller:0x00007fabe5156a38 @ruby_version="2_4_2", @ruby_source_folder="/Users/kyrylo/.pry.d/ruby-2_4_2", @curl_cmd="curl --fail -L https://github.com/ruby/ruby/archive/v2_4_2.tar.gz | tar xzvf - 2> /dev/null", @etag_binary="etags", @etag_cmd="find . -type f -name '*.[chy]' | etags - --no-members">).install(*(any args))
           expected: 1 time with any arguments
           received: 0 times with any arguments
     # ./spec/pry-doc_spec.rb:27:in `block (4 levels) in <top (required)>'

Finished in 0.21216 seconds (files took 0.82188 seconds to load)
40 examples, 1 failure
```

Any ideas?

**Comment by kyrylo** [comment]
I'm on MacOS.

**Comment by banister** [comment]
@kyrylo it was due to data persisting between spec examples....i fixed it by reducing the amount of state i store on the class, by extracting out a new class `SymbolMapBuilder` and then resetting `symbol_map` to nil before each example

https://github.com/pry/pry-doc/pull/85/commits/22bc8da6ccf3f560ea6558ca250e390ba614b8ff

**Comment by banister** [comment]
@kyrylo thanks....i tried to make `CFile` easier to understand here https://github.com/pry/pry-doc/pull/85/commits/130fb72e2dbae4d261b01e59414f51a115ce1906

**Comment by banister** [comment]
renamed `SymbolMapBuilder` to `ETagParser` which is exactly what it is, also opens up possibilities of supporting other tag formats (such as ctags) in the future by swapping in a different parser
